extensions/vscode/src/utils/vscode.ts

```diff
@@ -0,0 +1,109 @@
+// Copyright (C) 2024 by Posit Software, PBC.
+
+import { Uri, commands, workspace } from "vscode";
+import { fileExists, isDir } from "./files";
+import { delay } from "./throttle";
+import { substituteVariables } from "./variables";
+import { LanguageRuntimeMetadata, PositronApi } from "positron";
+
+export async function getPythonInterpreterPath(): Promise<string | undefined> {
+  const workspaceFolder = workspace.workspaceFolders?.[0];
+  if (workspaceFolder === undefined) {
+    return undefined;
+  }
+  let configuredPython: string | undefined;
+  try {
+    configuredPython = await commands.executeCommand<string>(
+      "python.interpreterPath",
+      { workspaceFolder: workspaceFolder },
+    );
+  } catch (error: unknown) {
+    console.error(
+      "getPythonInterpreterPath was unable to execute command. Error = ",
+      error,
+    );
+  }
+  if (configuredPython === undefined) {
+    return undefined;
+  }
+  let python = substituteVariables(configuredPython, true);
+  const pythonUri = Uri.file(python);
+
+  if (await isDir(pythonUri)) {
+    // Configured python can be a directory such as a virtual environment.
+    const names = [
+      "bin/python",
+      "bin/python3",
+      "Scripts/python.exe",
+      "Scripts/python3.exe",
+    ];
+    for (const name of names) {
+      const candidate = Uri.joinPath(pythonUri, name);
+      if (await fileExists(candidate)) {
+        python = candidate.fsPath;
+      }
+    }
+  }
+  console.log("Python interpreter path:", python);
+  return python;
+}
+
+declare global {
+  function acquirePositronApi(): PositronApi;
+}
+
+let positronApi: PositronApi | null | undefined;
+
+function getPositronApi(): PositronApi | null {
+  if (positronApi === undefined) {
+    try {
+      positronApi = acquirePositronApi();
+    } catch {
+      positronApi = null;
+    }
+  }
+  return positronApi;
+}
+
+export async function getRInterpreterPath(): Promise<string | undefined> {
+  const api = getPositronApi();
+
+  if (api) {
+    let runtime: LanguageRuntimeMetadata | undefined;
+
+    // Small number of retries, because getPreferredRuntime
+    // has its own internal retry logic.
+    const retries = 3;
+    const retryInterval = 1000;
+
+    for (let i = 0; i < retries + 1; i++) {
+      try {
+        runtime = await api.runtime.getPreferredRuntime("r");
```

Comment:
Just curious, if `getPreferredRuntime` has it's own retry logic, is it worth having that extra check again here? In which cases could it fail and then succeed?

Reply:
Not sure of Mikes' intentions. I'll take a look.

Reply:
`getPreferredRuntime` is an API that executes within Positron. They retry to handle queries during the initialization of Positron when the info is not yet available. However, there are timing issues in which the Positron extension API is not yet available, so we need to retry for it to become available. So we wait a bit for the API to become available and they wait until they have the initialization completed to know how to respond.

I don't think any change is required here.